compiler+runtime/src/cpp/jank/aot.cpp

```diff
@@ -0,0 +1,281 @@
+           void (*fn)(int const argc, char const **argv));
+}
+
+    )");
+
+    sb(R"(
+extern "C" jank_object_ref jank_var_intern(jank_object_ref ns, jank_object_ref var);
+extern "C" jank_object_ref jank_call0(jank_object_ref f);
+    )");
+
+    sb(R"(
+extern "C" jank_object_ref jank_load_clojure_core_native();
+extern "C" jank_object_ref jank_load_clojure_string_native();
+extern "C" jank_object_ref jank_load_jank_compiler_native();
+extern "C" jank_object_ref jank_load_clojure_core();
+extern "C" jank_object_ref jank_var_intern_c(char const *, char const *);
+extern "C" jank_object_ref jank_deref(jank_object_ref);
+extern "C" jank_object_ref jank_call0(jank_object_ref);
+)");
+
+    for(auto const &m : __rt_ctx->loaded_modules_in_order)
+    {
+      sb(R"(extern "C" jank_object_ref )");
+      sb(module::module_to_load_function(m));
+      sb(R"(();)");
+      sb("\n");
+    }
+
+    sb(R"(
+
+int main(int argc, const char** argv)
+{ // main start
+
+  using namespace jank;
+
+  auto const fn{ [](int const argc, char const **argv) {
+    jank_load_clojure_core_native();
+    jank_load_clojure_string_native();
+    jank_load_jank_compiler_native();
+
+    )");
+
+    for(auto const &m : __rt_ctx->loaded_modules_in_order)
+    {
+      sb(module::module_to_load_function(m));
+      sb(R"(();
+      )");
+    }
+
+    sb(R"(
+
+    // project modules loaded
+
+    // main function being called
+    
+    auto const fn(jank_var_intern_c(")");
+
+    sb(module);
+    sb(R"(", "-main"));
+    jank_call0(jank_deref(fn));
+
+  } };
+
+  return jank::init(argc, argv, /* init_default_ctx= */ true, fn);
+}
+  )");
+
+    auto const main_file(relative_to_cache_dir("/jank_main.cpp"));
+
+    std::ofstream out(main_file);
+    out << sb.release();
+
+    out.flush();
+    out.close();
+  }
+
+  void aot_compiler::compile(jtl::immutable_string const &module) const
+  {
+    gen_entrypoint(module);
+
+    llvm::IntrusiveRefCntPtr<clang::DiagnosticOptions> DiagOpts = new clang::DiagnosticOptions();
+
+    auto *DiagClient = new clang::TextDiagnosticPrinter(llvm::errs(), &*DiagOpts);
+
+    llvm::IntrusiveRefCntPtr<clang::DiagnosticIDs> DiagID(new clang::DiagnosticIDs());
+
+    llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> VFS = llvm::vfs::getRealFileSystem();
+
+    clang::DiagnosticsEngine Diags(DiagID, &*DiagOpts, DiagClient, /*ShouldOwnClient=*/true);
+
+
+    std::string const TargetTripleStr = llvm::sys::getDefaultTargetTriple();
+    llvm::errs() << "Using target triple: " << TargetTripleStr << "\n";
+
+    /* The Driver needs the path to the executable (used for finding related tools/resources)
+     * For this example, we'll use argv[0] or a placeholder. This might need adjustment
+     * depending on how/where your program is run relative to clang resources.
+     * TODO: infer clang++ path. Will be OS dependent and will require to infer correct version's path */
+    auto const ClangExecutablePath{ "/usr/bin/clang++" };
```

Comment:
The best way to infer clang's path is to just use `clang++` and rely on the user having set up their `PATH` the way they want. Guarantee you that `clang++` on my machine is not in `/usr/bin`. For Linux, it's best to never assume that something is in `/usr/bin`.

Reply:
Just using `clang++` doesn't work. I'm not sure if this way of doing compilation looks for the `clang++` binary. Instead, it expects the executable as the first arg.

The error I got was: `clang++: command not found`

Reply:
We can accept a CXX variable from command line for clang's executable path?

Although, ideally we should infer the path ourselves. We can try figuring it out by looking into known system installation directories and clang's version to make sure that we get the correct clang++ executable. But this won't always yield results. You talked about AppImage that will include the clang's executable. Before that, I think there's no deterministic way to get clang executable path.

Reply:
> Just using `clang++` doesn't work. I'm not sure if this way of doing compilation looks for the `clang++` binary. Instead, it expects the executable as the first arg.
> 
> The error I got was: `clang++: command not found`

Is `clang++` on your `PATH`? i.e. can you type it into a shell? Or do you only put it into `CXX`? We can look at `CXX`, but we can't rely on it, since an AOT compiled jank program shipped to some non-devs machine will not likely have `CXX` set. We need them to have clang installed, though, and we need to be able to find it. Relying on `PATH` is the most portable way, but we may also need more beyond that, such as looking into common places if we don't find it in `CXX` or `PATH`.